File: DS_A/C++/Dynamic_Programming/coin_repeatable_NOC.cpp
#include <iostream>
#include <limits>
#include <vector>
 
#define ll long long
 
using namespace std;
 
int main(){
	int n,x;
	cin>>n>>x;
	vector <int> tangalar(n+1); 
	
	int dump;
	for(int i = 0; i<n; i++){
		cin>>dump;
		tangalar[i]=dump;
	}
	
	vector <int> baza(x+1,numeric_limits<int>::max()); // baza yaratdik
	
	baza[0] = 1; // bazani sozladik
	
	for(int c: tangalar){ // 1-sikl! Tangalardan bir tanga tanlab oldik
		for(int q = c; q<=x; q++){ // 2-sikl! Tanlab olingan tangani hosil qilishimiz kerak bo'lgan qiymatga qarab surdik!
			baza[q]=min(baza[q],baza[q-c]+1); // solishtiryapmiz
		}
	}
	
	if(baza[x-1]==numeric_limits<int>::max()){ // numeric_limits<int>::max() bu maksimal son
		cout<<-1;
	}else{
		cout<<baza[x-1];
	}

File: DS_A/Python/Dynamic_Programming/coin_repeatable_NOC.py
# NOC -- Number of Coins (Minimal Tangalar Soni)

n,x = map(int,input().split()) #n tangalar soni, x hosil qilishimiz kerak bo'lgan qiymat'
tangalar = list(map(int,input().split()))

baza = [float('inf')]*(x+1) #baza yaratyapmiz
baza[0] = 0 # bazani sozlayapmiz

for c in tangalar: # c bu yerda 1-darjasi qism tanga
	for q in range(c,x+1): #q bu yerda 2-darajasli qism tanga biz 1-darajali tangadan hosil qilishimiz kerak bo'lgan tangaga qarab boramiz
		baza[q]=min(baza[q],baza[q-c]+1) # tavsiya etamiz hozirgi qiymat minimalmi yoki biz taklif qilgan tanga minimalmi? +1 bu qo'shilyotgan tanga soni baza[x-c] esa shu yacheykadagi tanga soni
	
	
if baza[x] == float('inf'):
	print(-1)
else:
	print(baza[x])
